src/buffer.c

```diff
@@ -455,6 +455,37 @@ int git_buf_decode_base85(
 	return -1;
 }
 
+#define HEX_DECODE(c) ((c | 32) % 39 - 9)
+
+int git_buf_decode_percent(
+	git_buf *buf,
+	const char *str,
+	size_t str_len)
+{
+	size_t str_pos, new_size;
+	int error = 0;
+
+	GITERR_CHECK_ALLOC_ADD(&new_size, buf->size, str_len);
+	GITERR_CHECK_ALLOC_ADD(&new_size, new_size, 1);
+	ENSURE_SIZE(buf, new_size);
+
+	for (str_pos = 0; str_pos < str_len; buf->size++, str_pos++) {
+		if (str[str_pos] == '%' &&
+			str_len > str_pos + 2 &&
```

Comment:
Hm. Doesn't this need to be `>=`? I'd expect like this if there is a "%xx" at the end of string it won't be decoded.

Reply:
I don't think so: consider some string of length 4 (`str_len` in this case is `4`).  If we were to use `>=` instead - then this comparison becomes `str_len >= str_pos + 2`, and then this would be true for `str_pos = 2`.  Consider that the string of length 4 is the string: `ab%z`.  At the point in the loop when `str_pos = 2`, then we would evaluate true on the first condition (`str[str_pos] == '%'`), and the second (`str_len >= str_pos + 2`), and then would walk off the end of the array in the fourth.  (`str[str_pos + 2]`).

Reply:
Agreed

Reply:
Convinced, thanks!